## 데이터 셋
영화 좋아하시나요?
- https://www.kaggle.com/datasets/harshitshankhdhar/imdb-dataset-of-top-1000-movies-and-tv-shows/

[별첨] # 마크다운 테이블 왼쪽 정렬을 위한 팁

In [6]:
%%html
<style> .rendered_html table{margin-left:5px} </style>

**데이터(data)**
이름 | 설명
- | -
Poster_Link | imdb가 사용하는 포스터 링크
Series_Title  | 영화 이름
Released_Year | 해당 영화가 개봉된 연도
Certificate | 인증서, 해당 영화로 얻은 인증서
Runtime  | 영화의 총 진행길이
Genre | 영화의 장르
IMDB_Rating | IMDB 사이트의 영화 등급
Overview | 개요. 미니스토리/요약
Meta_score | 영화에서 얻은 점수
Director | 영화 감독 이름
Star1,Star2,Star3,Star4 | 대표 배우 이름
No_of_votes | 총 투표 수
Gross | 해당 영화로 수익(돈)

**영감(Inspiration)**
- 영화 대 감독의 흥행 분석.
- 영화의 흥행 대 다른 스타의 총체적 분석.
- 영화 대 감독의 득표수 분석.
- 영화와 다른 스타의 투표수 분석.
- 어떤 배우가 어떤 장르를 더 선호하나요?
- IMDB_Rating 최대 시간이 좋은 배우 조합은 무엇입니까?
- 어떤 배우 조합이 좋은 성적을 내고 있나요?

## 실습

In [2]:
import pandas as pd

In [3]:
movie_df = pd.read_csv('imdb_top_1000.csv')
movie_df

Poster_Link  \
0    https://m.media-amazon.com/images/M/MV5BMDFkYT...   
1    https://m.media-amazon.com/images/M/MV5BM2MyNj...   
2    https://m.media-amazon.com/images/M/MV5BMTMxNT...   
3    https://m.media-amazon.com/images/M/MV5BMWMwMG...   
4    https://m.media-amazon.com/images/M/MV5BMWU4N2...   
..                                                 ...   
995  https://m.media-amazon.com/images/M/MV5BNGEwMT...   
996  https://m.media-amazon.com/images/M/MV5BODk3Yj...   
997  https://m.media-amazon.com/images/M/MV5BM2U3Yz...   
998  https://m.media-amazon.com/images/M/MV5BZTBmMj...   
999  https://m.media-amazon.com/images/M/MV5BMTY5OD...   

                 Series_Title Released_Year Certificate  Runtime  \
0    The Shawshank Redemption          1994           A  142 min   
1               The Godfather          1972           A  175 min   
2             The Dark Knight          2008          UA  152 min   
3      The Godfather: Part II          1974           A  202 min   
4                12 Angry Men          1957           U   96 min   
..                        ...           ...         ...      ...   
995    Breakfast at Tiffany's          1961           A  115 min   
996                     Giant          1956           G  201 min   
997     From Here to Eternity          1953      Passed  118 min   
998                  Lifeboat          1944         NaN   97 min   
999              The 39 Steps          1935         NaN   86 min   

                        Genre  IMDB_Rating  \
0                       Drama          9.3   
1                Crime, Drama          9.2   
2        Action, Crime, Drama          9.0   
3                Crime, Drama          9.0   
4                Crime, Drama          9.0   
..                        ...          ...   
995    Comedy, Drama, Romance          7.6   
996            Drama, Western          7.6   
997       Drama, Romance, War          7.6   
998                Drama, War          7.6   
999  Crime, Mystery, Thriller          7.6   

                                              Overview  Meta_score  \
0    Two imprisoned men bond over a number of years...        80.0   
1    An organized crime dynasty's aging patriarch t...       100.0   
2    When the menace known as the Joker wreaks havo...        84.0   
3    The early life and career of Vito Corleone in ...        90.0   
4    A jury holdout attempts to prevent a miscarria...        96.0   
..                                                 ...         ...   
995  A young New York socialite becomes interested ...        76.0   
996  Sprawling epic covering the life of a Texas ca...        84.0   
997  In Hawaii in 1941, a private is cruelly punish...        85.0   
998  Several survivors of a torpedoed merchant ship...        78.0   
999  A man in London tries to help a counter-espion...        93.0   

                 Director              Star1              Star2  \
0          Frank Darabont        Tim Robbins     Morgan Freeman   
1    Francis Ford Coppola      Marlon Brando          Al Pacino   
2       Christopher Nolan     Christian Bale       Heath Ledger   
3    Francis Ford Coppola          Al Pacino     Robert De Niro   
4            Sidney Lumet        Henry Fonda        Lee J. Cobb   
..                    ...                ...                ...   
995         Blake Edwards     Audrey Hepburn     George Peppard   
996        George Stevens   Elizabeth Taylor        Rock Hudson   
997        Fred Zinnemann     Burt Lancaster   Montgomery Clift   
998      Alfred Hitchcock  Tallulah Bankhead        John Hodiak   
999      Alfred Hitchcock       Robert Donat  Madeleine Carroll   

              Star3           Star4  No_of_Votes        Gross  
0        Bob Gunton  William Sadler      2343110   28,341,469  
1        James Caan    Diane Keaton      1620367  134,966,411  
2     Aaron Eckhart   Michael Caine      2303232  534,858,444  
3     Robert Duvall    Diane Keaton      1129952   57,300,000  
4     Mart

## 필요한 열만 가져오기

In [7]:
cols = ['Series_Title', 'Released_Year', 'Certificate', 'Genre', 'IMDB_Rating', 'Meta_score', 'Director', 'No_of_Votes', 'Gross']

In [8]:
# 가져올 열 지정하기
movie_df = pd.read_csv('imdb_top_1000.csv', usecols=cols)
movie_df

Series_Title Released_Year Certificate  \
0    The Shawshank Redemption          1994           A   
1               The Godfather          1972           A   
2             The Dark Knight          2008          UA   
3      The Godfather: Part II          1974           A   
4                12 Angry Men          1957           U   
..                        ...           ...         ...   
995    Breakfast at Tiffany's          1961           A   
996                     Giant          1956           G   
997     From Here to Eternity          1953      Passed   
998                  Lifeboat          1944         NaN   
999              The 39 Steps          1935         NaN   

                        Genre  IMDB_Rating  Meta_score              Director  \
0                       Drama          9.3        80.0        Frank Darabont   
1                Crime, Drama          9.2       100.0  Francis Ford Coppola   
2        Action, Crime, Drama          9.0        84.0     Christopher Nolan   
3                Crime, Drama          9.0        90.0  Francis Ford Coppola   
4                Crime, Drama          9.0        96.0          Sidney Lumet   
..                        ...          ...         ...                   ...   
995    Comedy, Drama, Romance          7.6        76.0         Blake Edwards   
996            Drama, Western          7.6        84.0        George Stevens   
997       Drama, Romance, War          7.6        85.0        Fred Zinnemann   
998                Drama, War          7.6        78.0      Alfred Hitchcock   
999  Crime, Mystery, Thriller          7.6        93.0      Alfred Hitchcock   

     No_of_Votes        Gross  
0        2343110   28,341,469  
1        1620367  134,966,411  
2        2303232  534,858,444  
3        1129952   57,300,000  
4         689845    4,360,000  
..           ...          ...  
995       166544          NaN  
996        34075          NaN  
997        43374   30,500,000  
998        26471          NaN  
999        51853          NaN  

[1000 rows x 9 columns]

## 열 순서를 변경하기
- 열의 우선순위를 생각해서 가져와보도록 하자.
- cols에 저장된 순서를 바꾸면 될까?

In [9]:
cols = ['Released_Year', 'Genre', 'Series_Title', 'Director', 'Meta_score', 'IMDB_Rating', 'No_of_Votes', 'Certificate', 'Gross']

In [10]:
# 순서 바꿔서 열 지정하기
movie_df = pd.read_csv('imdb_top_1000.csv', usecols=cols)
movie_df

Series_Title Released_Year Certificate  \
0    The Shawshank Redemption          1994           A   
1               The Godfather          1972           A   
2             The Dark Knight          2008          UA   
3      The Godfather: Part II          1974           A   
4                12 Angry Men          1957           U   
..                        ...           ...         ...   
995    Breakfast at Tiffany's          1961           A   
996                     Giant          1956           G   
997     From Here to Eternity          1953      Passed   
998                  Lifeboat          1944         NaN   
999              The 39 Steps          1935         NaN   

                        Genre  IMDB_Rating  Meta_score              Director  \
0                       Drama          9.3        80.0        Frank Darabont   
1                Crime, Drama          9.2       100.0  Francis Ford Coppola   
2        Action, Crime, Drama          9.0        84.0     Christopher Nolan   
3                Crime, Drama          9.0        90.0  Francis Ford Coppola   
4                Crime, Drama          9.0        96.0          Sidney Lumet   
..                        ...          ...         ...                   ...   
995    Comedy, Drama, Romance          7.6        76.0         Blake Edwards   
996            Drama, Western          7.6        84.0        George Stevens   
997       Drama, Romance, War          7.6        85.0        Fred Zinnemann   
998                Drama, War          7.6        78.0      Alfred Hitchcock   
999  Crime, Mystery, Thriller          7.6        93.0      Alfred Hitchcock   

     No_of_Votes        Gross  
0        2343110   28,341,469  
1        1620367  134,966,411  
2        2303232  534,858,444  
3        1129952   57,300,000  
4         689845    4,360,000  
..           ...          ...  
995       166544          NaN  
996        34075          NaN  
997        43374   30,500,000  
998        26471          NaN  
999        51853          NaN  

[1000 rows x 9 columns]

변경이 되지 않는다.
다음과 같은 방법으로 변경할 수 있다.

In [11]:
movie_df[cols]

Released_Year                     Genre              Series_Title  \
0            1994                     Drama  The Shawshank Redemption   
1            1972              Crime, Drama             The Godfather   
2            2008      Action, Crime, Drama           The Dark Knight   
3            1974              Crime, Drama    The Godfather: Part II   
4            1957              Crime, Drama              12 Angry Men   
..            ...                       ...                       ...   
995          1961    Comedy, Drama, Romance    Breakfast at Tiffany's   
996          1956            Drama, Western                     Giant   
997          1953       Drama, Romance, War     From Here to Eternity   
998          1944                Drama, War                  Lifeboat   
999          1935  Crime, Mystery, Thriller              The 39 Steps   

                 Director  Meta_score  IMDB_Rating  No_of_Votes Certificate  \
0          Frank Darabont        80.0          9.3      2343110           A   
1    Francis Ford Coppola       100.0          9.2      1620367           A   
2       Christopher Nolan        84.0          9.0      2303232          UA   
3    Francis Ford Coppola        90.0          9.0      1129952           A   
4            Sidney Lumet        96.0          9.0       689845           U   
..                    ...         ...          ...          ...         ...   
995         Blake Edwards        76.0          7.6       166544           A   
996        George Stevens        84.0          7.6        34075           G   
997        Fred Zinnemann        85.0          7.6        43374      Passed   
998      Alfred Hitchcock        78.0          7.6        26471         NaN   
999      Alfred Hitchcock        93.0          7.6        51853         NaN   

           Gross  
0     28,341,469  
1    134,966,411  
2    534,858,444  
3     57,300,000  
4      4,360,000  
..           ...  
995          NaN  
996          NaN  
997   30,500,000  
998          NaN  
999          NaN  

[1000 rows x 9 columns]

In [14]:
movie_df = movie_df[cols]
movie_df

Released_Year                     Genre              Series_Title  \
0            1994                     Drama  The Shawshank Redemption   
1            1972              Crime, Drama             The Godfather   
2            2008      Action, Crime, Drama           The Dark Knight   
3            1974              Crime, Drama    The Godfather: Part II   
4            1957              Crime, Drama              12 Angry Men   
..            ...                       ...                       ...   
995          1961    Comedy, Drama, Romance    Breakfast at Tiffany's   
996          1956            Drama, Western                     Giant   
997          1953       Drama, Romance, War     From Here to Eternity   
998          1944                Drama, War                  Lifeboat   
999          1935  Crime, Mystery, Thriller              The 39 Steps   

                 Director  Meta_score  IMDB_Rating  No_of_Votes Certificate  \
0          Frank Darabont        80.0          9.3      2343110           A   
1    Francis Ford Coppola       100.0          9.2      1620367           A   
2       Christopher Nolan        84.0          9.0      2303232          UA   
3    Francis Ford Coppola        90.0          9.0      1129952           A   
4            Sidney Lumet        96.0          9.0       689845           U   
..                    ...         ...          ...          ...         ...   
995         Blake Edwards        76.0          7.6       166544           A   
996        George Stevens        84.0          7.6        34075           G   
997        Fred Zinnemann        85.0          7.6        43374      Passed   
998      Alfred Hitchcock        78.0          7.6        26471         NaN   
999      Alfred Hitchcock        93.0          7.6        51853         NaN   

           Gross  
0     28,341,469  
1    134,966,411  
2    534,858,444  
3     57,300,000  
4      4,360,000  
..           ...  
995          NaN  
996          NaN  
997   30,500,000  
998          NaN  
999          NaN  

[1000 rows x 9 columns]

## .info()
- 간략한 요약 출력
- index dtype, columns, non-nulll values, memry usage

In [16]:
movie_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 9 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Released_Year  1000 non-null   object 
 1   Genre          1000 non-null   object 
 2   Series_Title   1000 non-null   object 
 3   Director       1000 non-null   object 
 4   Meta_score     843 non-null    float64
 5   IMDB_Rating    1000 non-null   float64
 6   No_of_Votes    1000 non-null   int64  
 7   Certificate    899 non-null    object 
 8   Gross          831 non-null    object 
dtypes: float64(2), int64(1), object(6)
memory usage: 70.4+ KB


**누락값(결측치) 존재 확인**
- 누락값이 Meta_score와 Certificate, Gross 열에서 보인다.
- 총 1000개의 데이터가 있어야하는데, `843 non-null`, `899 non-null`, `831 non-null`로 나온다.

## .describe()
- 집계 함수의 요약

In [17]:
movie_df.describe()

Meta_score  IMDB_Rating   No_of_Votes
count  843.000000  1000.000000  1.000000e+03
mean    77.971530     7.949300  2.736929e+05
std     12.376099     0.275491  3.273727e+05
min     28.000000     7.600000  2.508800e+04
25%     70.000000     7.700000  5.552625e+04
50%     79.000000     7.900000  1.385485e+05
75%     87.000000     8.100000  3.741612e+05
max    100.000000     9.300000  2.343110e+06

In [21]:
# 반올림 처리를 한다면
movie_df.describe().round()

Meta_score  IMDB_Rating  No_of_Votes
count       843.0       1000.0       1000.0
mean         78.0          8.0     273693.0
std          12.0          0.0     327373.0
min          28.0          8.0      25088.0
25%          70.0          8.0      55526.0
50%          79.0          8.0     138548.0
75%          87.0          8.0     374161.0
max         100.0          9.0    2343110.0

In [22]:
# 유효숫자 1자리 = 소수점 2자리에서 반올림 처리
movie_df.describe().round(1)

Meta_score  IMDB_Rating  No_of_Votes
count       843.0       1000.0       1000.0
mean         78.0          7.9     273692.9
std          12.4          0.3     327372.7
min          28.0          7.6      25088.0
25%          70.0          7.7      55526.2
50%          79.0          7.9     138548.5
75%          87.0          8.1     374161.2
max         100.0          9.3    2343110.0

## 왜 수익은 .describe()에서 안나올까? 

In [26]:
# Gross의 타입을 보면 숫자형 타입이 아닌 object임을 알 수 있다.
movie_df['Gross'].dtype

dtype('O')

In [27]:
# Gross의 포맷을 확인
movie_df.head()

Released_Year                 Genre              Series_Title  \
0          1994                 Drama  The Shawshank Redemption   
1          1972          Crime, Drama             The Godfather   
2          2008  Action, Crime, Drama           The Dark Knight   
3          1974          Crime, Drama    The Godfather: Part II   
4          1957          Crime, Drama              12 Angry Men   

               Director  Meta_score  IMDB_Rating  No_of_Votes Certificate  \
0        Frank Darabont        80.0          9.3      2343110           A   
1  Francis Ford Coppola       100.0          9.2      1620367           A   
2     Christopher Nolan        84.0          9.0      2303232          UA   
3  Francis Ford Coppola        90.0          9.0      1129952           A   
4          Sidney Lumet        96.0          9.0       689845           U   

         Gross  
0   28,341,469  
1  134,966,411  
2  534,858,444  
3   57,300,000  
4    4,360,000

## .str.replace() - 1단계 : 포맷 변경하기
- 숫자형 데이터로 변환 가능한 형태로 처리
- 데이터 분석에 유용한 형태로 가공하는 작업 : 데이터 전처리
- comma(,)부터 .str.replace() 로 삭제하자.
- https://pandas.pydata.org/docs/reference/api/pandas.Series.str.replace.html

In [33]:
# 1단계 comma(,) 삭제하기
movie_df['Gross']=movie_df['Gross'].str.replace(',','')

In [30]:
# 확인
movie_df

Released_Year                     Genre              Series_Title  \
0            1994                     Drama  The Shawshank Redemption   
1            1972              Crime, Drama             The Godfather   
2            2008      Action, Crime, Drama           The Dark Knight   
3            1974              Crime, Drama    The Godfather: Part II   
4            1957              Crime, Drama              12 Angry Men   
..            ...                       ...                       ...   
995          1961    Comedy, Drama, Romance    Breakfast at Tiffany's   
996          1956            Drama, Western                     Giant   
997          1953       Drama, Romance, War     From Here to Eternity   
998          1944                Drama, War                  Lifeboat   
999          1935  Crime, Mystery, Thriller              The 39 Steps   

                 Director  Meta_score  IMDB_Rating  No_of_Votes Certificate  \
0          Frank Darabont        80.0          9.3      2343110           A   
1    Francis Ford Coppola       100.0          9.2      1620367           A   
2       Christopher Nolan        84.0          9.0      2303232          UA   
3    Francis Ford Coppola        90.0          9.0      1129952           A   
4            Sidney Lumet        96.0          9.0       689845           U   
..                    ...         ...          ...          ...         ...   
995         Blake Edwards        76.0          7.6       166544           A   
996        George Stevens        84.0          7.6        34075           G   
997        Fred Zinnemann        85.0          7.6        43374      Passed   
998      Alfred Hitchcock        78.0          7.6        26471         NaN   
999      Alfred Hitchcock        93.0          7.6        51853         NaN   

         Gross  
0     28341469  
1    134966411  
2    534858444  
3     57300000  
4      4360000  
..         ...  
995        NaN  
996        NaN  
997   30500000  
998        NaN  
999        NaN  

[1000 rows x 9 columns]

## .fillna() - 2단계 : 결측치 대체하기
- NaN으로 나오고 있는 결측치는 0으로 대체

In [35]:
# movie_df['Gross'].fillna(0, inplace=True)
movie_df['Gross'] =movie_df['Gross'].fillna(0)
movie_df

Released_Year                     Genre              Series_Title  \
0            1994                     Drama  The Shawshank Redemption   
1            1972              Crime, Drama             The Godfather   
2            2008      Action, Crime, Drama           The Dark Knight   
3            1974              Crime, Drama    The Godfather: Part II   
4            1957              Crime, Drama              12 Angry Men   
..            ...                       ...                       ...   
995          1961    Comedy, Drama, Romance    Breakfast at Tiffany's   
996          1956            Drama, Western                     Giant   
997          1953       Drama, Romance, War     From Here to Eternity   
998          1944                Drama, War                  Lifeboat   
999          1935  Crime, Mystery, Thriller              The 39 Steps   

                 Director  Meta_score  IMDB_Rating  No_of_Votes Certificate  \
0          Frank Darabont        80.0          9.3      2343110           A   
1    Francis Ford Coppola       100.0          9.2      1620367           A   
2       Christopher Nolan        84.0          9.0      2303232          UA   
3    Francis Ford Coppola        90.0          9.0      1129952           A   
4            Sidney Lumet        96.0          9.0       689845           U   
..                    ...         ...          ...          ...         ...   
995         Blake Edwards        76.0          7.6       166544           A   
996        George Stevens        84.0          7.6        34075           G   
997        Fred Zinnemann        85.0          7.6        43374      Passed   
998      Alfred Hitchcock        78.0          7.6        26471         NaN   
999      Alfred Hitchcock        93.0          7.6        51853         NaN   

         Gross  
0     28341469  
1    134966411  
2    534858444  
3     57300000  
4      4360000  
..         ...  
995          0  
996          0  
997   30500000  
998          0  
999          0  

[1000 rows x 9 columns]

In [37]:
# .info()로도 .count()로도 결측치 확인이 가능
movie_df.info()
movie_df.count()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 9 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Released_Year  1000 non-null   object 
 1   Genre          1000 non-null   object 
 2   Series_Title   1000 non-null   object 
 3   Director       1000 non-null   object 
 4   Meta_score     843 non-null    float64
 5   IMDB_Rating    1000 non-null   float64
 6   No_of_Votes    1000 non-null   int64  
 7   Certificate    899 non-null    object 
 8   Gross          1000 non-null   object 
dtypes: float64(2), int64(1), object(6)
memory usage: 70.4+ KB


Released_Year    1000
Genre            1000
Series_Title     1000
Director         1000
Meta_score        843
IMDB_Rating      1000
No_of_Votes      1000
Certificate       899
Gross            1000
dtype: int64

## .astype() - 3단계 : dtype 변환하기
- 정수형 dtype인 int64로 변환

In [41]:
# .astype('변환형')
movie_df['Gross']=movie_df['Gross'].astype('int64')

In [42]:
# 데이터 조회하면
movie_df

Released_Year                     Genre              Series_Title  \
0            1994                     Drama  The Shawshank Redemption   
1            1972              Crime, Drama             The Godfather   
2            2008      Action, Crime, Drama           The Dark Knight   
3            1974              Crime, Drama    The Godfather: Part II   
4            1957              Crime, Drama              12 Angry Men   
..            ...                       ...                       ...   
995          1961    Comedy, Drama, Romance    Breakfast at Tiffany's   
996          1956            Drama, Western                     Giant   
997          1953       Drama, Romance, War     From Here to Eternity   
998          1944                Drama, War                  Lifeboat   
999          1935  Crime, Mystery, Thriller              The 39 Steps   

                 Director  Meta_score  IMDB_Rating  No_of_Votes Certificate  \
0          Frank Darabont        80.0          9.3      2343110           A   
1    Francis Ford Coppola       100.0          9.2      1620367           A   
2       Christopher Nolan        84.0          9.0      2303232          UA   
3    Francis Ford Coppola        90.0          9.0      1129952           A   
4            Sidney Lumet        96.0          9.0       689845           U   
..                    ...         ...          ...          ...         ...   
995         Blake Edwards        76.0          7.6       166544           A   
996        George Stevens        84.0          7.6        34075           G   
997        Fred Zinnemann        85.0          7.6        43374      Passed   
998      Alfred Hitchcock        78.0          7.6        26471         NaN   
999      Alfred Hitchcock        93.0          7.6        51853         NaN   

         Gross  
0     28341469  
1    134966411  
2    534858444  
3     57300000  
4      4360000  
..         ...  
995          0  
996          0  
997   30500000  
998          0  
999          0  

[1000 rows x 9 columns]

In [44]:
# .describe()로 요약하면
movie_df.describe()

Meta_score  IMDB_Rating   No_of_Votes         Gross
count  843.000000  1000.000000  1.000000e+03  1.000000e+03
mean    77.971530     7.949300  2.736929e+05  5.653688e+07
std     12.376099     0.275491  3.273727e+05  1.032382e+08
min     28.000000     7.600000  2.508800e+04  0.000000e+00
25%     70.000000     7.700000  5.552625e+04  4.457098e+05
50%     79.000000     7.900000  1.385485e+05  1.070275e+07
75%     87.000000     8.100000  3.741612e+05  6.153989e+07
max    100.000000     9.300000  2.343110e+06  9.366622e+08

## percentiles=
- 각 OO% 백분위수를 확인할 수 있다.

In [46]:
# 필요하다면 퍼센트 범위를 지정할 수도 있다.
# [20%, 40%, 60%, 80%] -> [.2, .4, .6, .8]
movie_df.describe(percentiles=[.2,.4,.6,.8]).round(1)

Meta_score  IMDB_Rating  No_of_Votes        Gross
count       843.0       1000.0       1000.0       1000.0
mean         78.0          7.9     273692.9   56536878.0
std          12.4          0.3     327372.7  103238179.5
min          28.0          7.6      25088.0          0.0
20%          68.0          7.7      48532.4      83937.8
40%          76.0          7.8      91409.4    4164721.6
50%          79.0          7.9     138548.5   10702751.5
60%          82.0          8.0     203747.8   25554220.6
80%          89.0          8.1     462439.2   85388198.4
max         100.0          9.3    2343110.0  936662225.0

>**percentiles : list-like of numbers, optional**  
The percentiles to include in the output.   
All should fall between 0 and 1. The default is ``[.25, .5, .75]``, which returns the 25th, 50th, and 75th percentiles.

In [47]:
# 빈 리스트를 넣으면? -> 50%는 남는다
movie_df.describe(percentiles=[]).round(1)

Meta_score  IMDB_Rating  No_of_Votes        Gross
count       843.0       1000.0       1000.0       1000.0
mean         78.0          7.9     273692.9   56536878.0
std          12.4          0.3     327372.7  103238179.5
min          28.0          7.6      25088.0          0.0
50%          79.0          7.9     138548.5   10702751.5
max         100.0          9.3    2343110.0  936662225.0

- 왜냐하면 50%는 사실은 median(중앙값)이고, 이 때문에 percentiles처리에 오류가 있다는 오해가 종종 있다.

## DataFrame 집계함수에서 numeric_only=
- max, min, mean 등을 DataFrame에서 바로 사용하면 때때로 경고(Warning)이 발생한다.
- 숫자형 데이터만 있는 건 아니기 때문임으로 숫자만 처리하도록 지정하는 것.

In [50]:
# 이러면 Warning 발생 - max
movie_df.max()

C:\Users\82103\AppData\Local\Temp\ipykernel_25528\2269447593.py:2: FutureWarning: Dropping of nuisance columns in DataFrame reductions (with 'numeric_only=None') is deprecated; in a future version this will raise TypeError.  Select only valid columns before calling the reduction.
  movie_df.max()


Released_Year                             PG
Genre                                Western
Series_Title     Ôkami kodomo no Ame to Yuki
Director                    Ömer Faruk Sorak
Meta_score                             100.0
IMDB_Rating                              9.3
No_of_Votes                          2343110
Gross                              936662225
dtype: object

In [52]:
# numeric_only=True
movie_df.max(numeric_only=True)

Meta_score           100.0
IMDB_Rating            9.3
No_of_Votes      2343110.0
Gross          936662225.0
dtype: float64

In [53]:
movie_df.min(numeric_only=True)

Meta_score        28.0
IMDB_Rating        7.6
No_of_Votes    25088.0
Gross              0.0
dtype: float64

In [54]:
movie_df.sum(numeric_only=True)

Meta_score     6.573000e+04
IMDB_Rating    7.949300e+03
No_of_Votes    2.736929e+08
Gross          5.653688e+10
dtype: float64

## .quantile()
- quantile : 분위수 - 오름차순/내림차순 되있을 때 이를 나누는 지점
  - percentiles : 백분위수, 100개 영역 지점이 존재. 퍼센트
  - quartiles : 4분위수, 4구역 지점이 존재

In [59]:
# 만일 50% 지점을 확인q한다면 0.5
movie_df.quantile(q=0.2, numeric_only=True)

Meta_score        68.0
IMDB_Rating        7.7
No_of_Votes    48532.4
Gross          83937.8
Name: 0.2, dtype: float64

In [62]:
# 50% 지점 == median(중앙값)
movie_df.median(numeric_only=True)

Meta_score           79.0
IMDB_Rating           7.9
No_of_Votes      138548.5
Gross          10702751.5
dtype: float64

## include=

> **include : ‘all’, list-like of dtypes or None (default), optional**  
A white list of data types to include in the result. Ignored for Series. Here are the options:  
  
- all : All columns of the input will be included in the output.  
- A list-like of dtypes : Limits the results to the provided data types. To limit the result to numeric types submit numpy.number. To limit it instead to object columns submit the numpy.object data type. Strings can also be used in the style of select_dtypes (e.g. **df.describe(include=['O'])**). To select pandas categorical columns, use 'category'  
- None (default) : The result will include all numeric columns.  

---
1. 'all': 입력된 모든 열이 결과에 포함됩니다.
2. 데이터 유형의 목록 형태(list-like of dtypes): 결과를 지정된 데이터 유형으로 제한합니다. 
    - 숫자 유형에 대한 결과를 제한하려면 numpy.number를 사용하고, 
    - 객체 열에 대한 결과를 제한하려면 numpy.object 데이터 유형을 사용합니다. 
    - select_dtypes와 유사한 방식으로 문자열도 사용할 수 있습니다 (예: **df.describe(include=['O'])**). 알파벳임. 0아님  
    - pandas 범주형 열을 선택하려면 'category'를 사용합니다.
3. None (기본값): 결과에는 모든 숫자 열이 포함됩니다.

In [71]:
movie_df.describe(include=['O'])

Released_Year  Genre Series_Title          Director Certificate
count           1000   1000         1000              1000         899
unique           100    202          999               548          16
top             2014  Drama     Drishyam  Alfred Hitchcock           U
freq              32     85            2                14         234

- count: 해당 열의 비어 있지 않은(non-null) 값의 개수
- unique: 해당 열에서 고유한(unique) 값의 개수
- top: 해당 열에서 가장 빈도가 높은 값(top value)
- freq: top 값의 빈도수

In [72]:
# 문자열 타입에 대한 요약 정보
movie_df.describe(include=['object'])

Released_Year  Genre Series_Title          Director Certificate
count           1000   1000         1000              1000         899
unique           100    202          999               548          16
top             2014  Drama     Drishyam  Alfred Hitchcock           U
freq              32     85            2                14         234

In [73]:
# 숫자형 타입에 대한 요약 정보
movie_df.describe(include=['int'])


No_of_Votes         Gross
count  1.000000e+03  1.000000e+03
mean   2.736929e+05  5.653688e+07
std    3.273727e+05  1.032382e+08
min    2.508800e+04  0.000000e+00
25%    5.552625e+04  4.457098e+05
50%    1.385485e+05  1.070275e+07
75%    3.741612e+05  6.153989e+07
max    2.343110e+06  9.366622e+08

In [75]:
# 문자열, int형에 대한 요약 정보
movie_df.describe(include=['O','int'])

Released_Year  Genre Series_Title          Director   No_of_Votes  \
count           1000   1000         1000              1000  1.000000e+03   
unique           100    202          999               548           NaN   
top             2014  Drama     Drishyam  Alfred Hitchcock           NaN   
freq              32     85            2                14           NaN   
mean             NaN    NaN          NaN               NaN  2.736929e+05   
std              NaN    NaN          NaN               NaN  3.273727e+05   
min              NaN    NaN          NaN               NaN  2.508800e+04   
25%              NaN    NaN          NaN               NaN  5.552625e+04   
50%              NaN    NaN          NaN               NaN  1.385485e+05   
75%              NaN    NaN          NaN               NaN  3.741612e+05   
max              NaN    NaN          NaN               NaN  2.343110e+06   

       Certificate         Gross  
count          899  1.000000e+03  
unique          16           NaN  
top              U           NaN  
freq           234           NaN  
mean           NaN  5.653688e+07  
std            NaN  1.032382e+08  
min            NaN  0.000000e+00  
25%            NaN  4.457098e+05  
50%            NaN  1.070275e+07  
75%            NaN  6.153989e+07  
max            NaN  9.366622e+08

In [77]:
# 모든(all) 타입에 대한 요약 정보
movie_df.describe(include='all')

Released_Year  Genre Series_Title          Director  Meta_score  \
count           1000   1000         1000              1000  843.000000   
unique           100    202          999               548         NaN   
top             2014  Drama     Drishyam  Alfred Hitchcock         NaN   
freq              32     85            2                14         NaN   
mean             NaN    NaN          NaN               NaN   77.971530   
std              NaN    NaN          NaN               NaN   12.376099   
min              NaN    NaN          NaN               NaN   28.000000   
25%              NaN    NaN          NaN               NaN   70.000000   
50%              NaN    NaN          NaN               NaN   79.000000   
75%              NaN    NaN          NaN               NaN   87.000000   
max              NaN    NaN          NaN               NaN  100.000000   

        IMDB_Rating   No_of_Votes Certificate         Gross  
count   1000.000000  1.000000e+03         899  1.000000e+03  
unique          NaN           NaN          16           NaN  
top             NaN           NaN           U           NaN  
freq            NaN           NaN         234           NaN  
mean       7.949300  2.736929e+05         NaN  5.653688e+07  
std        0.275491  3.273727e+05         NaN  1.032382e+08  
min        7.600000  2.508800e+04         NaN  0.000000e+00  
25%        7.700000  5.552625e+04         NaN  4.457098e+05  
50%        7.900000  1.385485e+05         NaN  1.070275e+07  
75%        8.100000  3.741612e+05         NaN  6.153989e+07  
max        9.300000  2.343110e+06         NaN  9.366622e+08

표현할 수 없는 항목은 NaN으로 나온다.
- 문자열 데이터는 mean, std, min등을 표현할 수 없고
- 숫자형 데이터는 unique, top, freq를 표현할 수 있을 것 같은데, 크게 의미가 없습니다.
  - unique: 숫자형 데이터는 고유한(unique) 값의 개수를 가질 수 없으므로 NaN으로 처리
  - top: 숫자형 데이터는 가장 빈도가 높은(top) 값이라는 개념이 적용되지 않으므로 NaN으로 처리
  - freq: 숫자형 데이터는 각 값의 빈도수를 측정할 수 없으므로 NaN으로 처리